### PR TITLE
ci: add Hyperfine pipelines

### LINKED ISSUE #95: Add hyperfine to the performance automated testing
https://github.com/sharkdp/hyperfine

### LINKED ISSUE #239: Post benchmark results on PR discussion
Currently our benchmark pipeline only posts comments when there's a significant performance regression. It would be useful for it to also post in all other cases, specially so we can confirm the effect of optimization PRs.

### PR DESCRIPTION
**Pull request by Oppen**
This replaces the IAI PR workflow with a parallel set of Hyperfine based benchmarks. It lowers the time spent in benchmarks from about 1h 10' to about 10'.
The workflow is definitely a complex one. It does the following steps:
- Build the benchmark programs in parallel with as many cores as available (2, in practice) for both the base and the HEAD commits.
- Build the `cairo-vm-cli` binary for both the base and the HEAD commits.
- Compare with `Hyperfine` the time spent by each binary running each of the benchmarks compiled from the base commit.
- Run exclusively whatever benchmarks were added or changed with the HEAD binary. The reason this is a separate step is because often new benchmarks are the consequence of new functionality not present in base, meaning it would fail the job.

Experiments on this PR show that:
- Hyperfine detects outliers sometimes.
- All results are, as expected, `1.00 +- 0.0x`, where `x in {0,1,2}`. This means any difference is attributed to statistical noise.
Experiments are included as comments. Only the more systematic ones are included, some prior ones were used for exploration.

The IAI benchmark is only disabled for PRs as it will come back into use for microbenchmarks.

Fixes #239 
Fixes #95 

Pending tasks:
- [x] Picking better names for the jobs and steps

## Checklist
- [ ] Linked to Github Issue
- [ ] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
  - [ ] CHANGELOG has been updated.

### CODE REVIEW

**Comment by Oppen** [comment]
Experiment: running on a single core, 3 warmup rounds, no prepare or setup stage.
The following runs had outliers:
- base linear_search
- head pedersen

Reported results:
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base pedersen` | 3.265 ± 0.093 | 3.197 | 3.409 | 1.02 ± 0.03 |
| `head pedersen` | 3.209 ± 0.046 | 3.190 | 3.340 | 1.00 |
base_programs/dict_integration_benchmark.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base dict_integration_benchmark` | 2.425 ± 0.011 | 2.406 | 2.440 | 1.00 ± 0.01 |
| `head dict_integration_benchmark` | 2.417 ± 0.013 | 2.399 | 2.441 | 1.00 |
base_programs/fibonacci_1000_multirun.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base fibonacci_1000_multirun` | 3.057 ± 0.020 | 3.037 | 3.104 | 1.01 ± 0.01 |
| `head fibonacci_1000_multirun` | 3.025 ± 0.010 | 3.012 | 3.040 | 1.00 |
base_programs/set_integration_benchmark.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base set_integration_benchmark` | 1.690 ± 0.024 | 1.674 | 1.755 | 1.00 |
| `head set_integration_benchmark` | 1.691 ± 0.007 | 1.682 | 1.702 | 1.00 ± 0.01 |
base_programs/integration_builtins.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base integration_builtins` | 3.074 ± 0.010 | 3.056 | 3.089 | 1.00 |
| `head integration_builtins` | 3.090 ± 0.014 | 3.064 | 3.108 | 1.00 ± 0.01 |
base_programs/factorial_multirun.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base factorial_multirun` | 3.828 ± 0.022 | 3.809 | 3.871 | 1.00 |
| `head factorial_multirun` | 3.830 ± 0.023 | 3.807 | 3.870 | 1.00 ± 0.01 |
base_programs/keccak_integration_benchmark.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base keccak_integration_benchmark` | 2.709 ± 0.023 | 2.691 | 2.769 | 1.01 ± 0.01 |
| `head keccak_integration_benchmark` | 2.673 ± 0.007 | 2.658 | 2.681 | 1.00 |
base_programs/poseidon_integration_benchmark.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base poseidon_integration_benchmark` | 1.504 ± 0.003 | 1.500 | 1.509 | 1.00 |
| `head poseidon_integration_benchmark` | 1.507 ± 0.002 | 1.504 | 1.510 | 1.00 ± 0.00 |
base_programs/linear_search.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base linear_search` | 3.224 ± 0.052 | 3.192 | 3.371 | 1.01 ± 0.02 |
| `head linear_search` | 3.195 ± 0.013 | 3.178 | 3.215 | 1.00 |
base_programs/blake2s_integration_benchmark.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base blake2s_integration_benchmark` | 2.508 ± 0.007 | 2.494 | 2.516 | 1.00 ± 0.01 |
| `head blake2s_integration_benchmark` | 2.505 ± 0.011 | 2.482 | 2.522 | 1.00 |
base_programs/math_cmp_and_pow_integration_benchmark.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base math_cmp_and_pow_integration_benchmark` | 2.735 ± 0.004 | 2.730 | 2.742 | 1.01 ± 0.01 |
| `head math_cmp_and_pow_integration_benchmark` | 2.710 ± 0.017 | 2.693 | 2.752 | 1.00 |
base_programs/operations_with_data_structures_benchmarks.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base operations_with_data_structures_benchmarks` | 2.241 ± 0.012 | 2.230 | 2.271 | 1.00 |
| `head operations_with_data_structures_benchmarks` | 2.253 ± 0.010 | 2.238 | 2.272 | 1.01 ± 0.01 |
base_programs/secp_integration_benchmark.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base secp_integration_benchmark` | 2.353 ± 0.007 | 2.344 | 2.365 | 1.00 |
| `head secp_integration_benchmark` | 2.364 ± 0.007 | 2.356 | 2.375 | 1.00 ± 0.00 |
base_programs/math_integration_benchmark.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base math_integration_benchmark` | 2.518 ± 0.012 | 2.503 | 2.538 | 1.00 |
| `head math_integration_benchmark` | 2.530 ± 0.005 | 2.521 | 2.537 | 1.00 ± 0.01 |
base_programs/compare_arrays_200000.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base compare_arrays_200000` | 3.079 ± 0.010 | 3.065 | 3.092 | 1.00 |
| `head compare_arrays_200000` | 3.096 ± 0.015 | 3.079 | 3.126 | 1.01 ± 0.01 |
base_programs/uint256_integration_benchmark.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base uint256_integration_benchmark` | 3.939 ± 0.011 | 3.920 | 3.952 | 1.00 ± 0.01 |
| `head uint256_integration_benchmark` | 3.924 ± 0.051 | 3.886 | 4.062 | 1.00 |
base_programs/memory_integration_benchmark.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base memory_integration_benchmark` | 2.159 ± 0.005 | 2.153 | 2.166 | 1.00 ± 0.00 |
| `head memory_integration_benchmark` | 2.150 ± 0.006 | 2.143 | 2.158 | 1.00 |

**Comment by Oppen** [comment]
Experiment 2: running on two cores, 3 warmup rounds, no prepare or setup stage.
It spent 13'37" in the run benchmarks step. The following runs had outliers:
- base math_integration_benchmark (first benchmark slower than the rest by about 10%)
- head memory_integration_benchmark
- base operations_with_data_structures_benchmarks
- base set_integration_benchmark

Reported results:
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base blake2s_integration_benchmark` | 3.259 ± 0.012 | 3.242 | 3.284 | 1.00 |
| `head blake2s_integration_benchmark` | 3.260 ± 0.060 | 3.224 | 3.423 | 1.00 ± 0.02 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base compare_arrays_200000` | 3.928 ± 0.049 | 3.889 | 4.047 | 1.00 |
| `head compare_arrays_200000` | 3.945 ± 0.026 | 3.914 | 4.005 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base dict_integration_benchmark` | 2.978 ± 0.015 | 2.960 | 3.001 | 1.00 ± 0.01 |
| `head dict_integration_benchmark` | 2.973 ± 0.015 | 2.954 | 2.999 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base factorial_multirun` | 4.818 ± 0.017 | 4.790 | 4.852 | 1.00 |
| `head factorial_multirun` | 4.819 ± 0.037 | 4.788 | 4.899 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base fibonacci_1000_multirun` | 3.830 ± 0.026 | 3.803 | 3.876 | 1.00 |
| `head fibonacci_1000_multirun` | 3.839 ± 0.047 | 3.800 | 3.954 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base integration_builtins` | 3.927 ± 0.029 | 3.898 | 3.995 | 1.00 |
| `head integration_builtins` | 3.935 ± 0.031 | 3.906 | 3.997 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base keccak_integration_benchmark` | 3.551 ± 0.011 | 3.535 | 3.576 | 1.00 |
| `head keccak_integration_benchmark` | 3.570 ± 0.051 | 3.531 | 3.700 | 1.01 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base linear_search` | 4.093 ± 0.040 | 4.049 | 4.188 | 1.00 ± 0.01 |
| `head linear_search` | 4.080 ± 0.025 | 4.047 | 4.131 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base math_cmp_and_pow_integration_benchmark` | 3.429 ± 0.018 | 3.403 | 3.462 | 1.00 |
| `head math_cmp_and_pow_integration_benchmark` | 3.471 ± 0.091 | 3.406 | 3.667 | 1.01 ± 0.03 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base math_integration_benchmark` | 3.108 ± 0.009 | 3.097 | 3.129 | 1.00 |
| `head math_integration_benchmark` | 3.210 ± 0.265 | 3.090 | 3.943 | 1.03 ± 0.09 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base memory_integration_benchmark` | 2.751 ± 0.014 | 2.729 | 2.773 | 1.00 |
| `head memory_integration_benchmark` | 2.830 ± 0.250 | 2.725 | 3.540 | 1.03 ± 0.09 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base operations_with_data_structures_benchmarks` | 2.772 ± 0.009 | 2.760 | 2.785 | 1.00 ± 0.01 |
| `head operations_with_data_structures_benchmarks` | 2.766 ± 0.013 | 2.750 | 2.783 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base pedersen` | 4.001 ± 0.012 | 3.987 | 4.023 | 1.00 |
| `head pedersen` | 4.001 ± 0.016 | 3.978 | 4.029 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base poseidon_integration_benchmark` | 1.754 ± 0.005 | 1.747 | 1.764 | 1.00 |
| `head poseidon_integration_benchmark` | 1.765 ± 0.005 | 1.756 | 1.770 | 1.01 ± 0.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base secp_integration_benchmark` | 3.103 ± 0.012 | 3.091 | 3.124 | 1.00 |
| `head secp_integration_benchmark` | 3.113 ± 0.010 | 3.101 | 3.136 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base set_integration_benchmark` | 2.175 ± 0.030 | 2.156 | 2.257 | 1.00 ± 0.01 |
| `head set_integration_benchmark` | 2.164 ± 0.004 | 2.161 | 2.174 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base uint256_integration_benchmark` | 4.857 ± 0.020 | 4.829 | 4.890 | 1.00 |
| `head uint256_integration_benchmark` | 4.863 ± 0.012 | 4.843 | 4.879 | 1.00 ± 0.00 |

**Comment by Oppen** [comment]
Experiment 3: running on two cores, 3 warmup rounds, no prepare or setup stage.
It spent 11'37" in the run benchmarks step. The following runs had outliers:
- base keccak_integration_benchmark
- base operations_with_data_structures_benchmarks
NOTE: due to interleaved output there may be mistakes about __which__ benchmarks really had the outliers, but the number of them that had is correct.

Reported results were:
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base blake2s_integration_benchmark` | 2.783 ± 0.035 | 2.760 | 2.872 | 1.01 ± 0.01 |
| `head blake2s_integration_benchmark` | 2.764 ± 0.009 | 2.753 | 2.776 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base compare_arrays_200000` | 3.386 ± 0.037 | 3.357 | 3.466 | 1.01 ± 0.01 |
| `head compare_arrays_200000` | 3.363 ± 0.014 | 3.334 | 3.379 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base dict_integration_benchmark` | 2.527 ± 0.011 | 2.515 | 2.553 | 1.00 ± 0.01 |
| `head dict_integration_benchmark` | 2.526 ± 0.015 | 2.508 | 2.558 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base factorial_multirun` | 4.118 ± 0.015 | 4.099 | 4.142 | 1.00 |
| `head factorial_multirun` | 4.130 ± 0.018 | 4.102 | 4.161 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base fibonacci_1000_multirun` | 3.286 ± 0.031 | 3.258 | 3.367 | 1.00 |
| `head fibonacci_1000_multirun` | 3.307 ± 0.040 | 3.256 | 3.406 | 1.01 ± 0.02 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base integration_builtins` | 3.368 ± 0.050 | 3.342 | 3.507 | 1.00 |
| `head integration_builtins` | 3.371 ± 0.028 | 3.345 | 3.441 | 1.00 ± 0.02 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base keccak_integration_benchmark` | 3.076 ± 0.029 | 3.049 | 3.150 | 1.00 ± 0.01 |
| `head keccak_integration_benchmark` | 3.065 ± 0.006 | 3.056 | 3.074 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base linear_search` | 3.480 ± 0.014 | 3.457 | 3.500 | 1.00 |
| `head linear_search` | 3.482 ± 0.021 | 3.452 | 3.530 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base math_cmp_and_pow_integration_benchmark` | 2.945 ± 0.012 | 2.925 | 2.965 | 1.00 |
| `head math_cmp_and_pow_integration_benchmark` | 2.950 ± 0.019 | 2.929 | 2.988 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base math_integration_benchmark` | 2.657 ± 0.040 | 2.632 | 2.769 | 1.00 |
| `head math_integration_benchmark` | 2.664 ± 0.065 | 2.634 | 2.845 | 1.00 ± 0.03 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base memory_integration_benchmark` | 2.352 ± 0.018 | 2.325 | 2.394 | 1.00 ± 0.01 |
| `head memory_integration_benchmark` | 2.350 ± 0.013 | 2.324 | 2.366 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base operations_with_data_structures_benchmarks` | 2.371 ± 0.007 | 2.360 | 2.380 | 1.00 |
| `head operations_with_data_structures_benchmarks` | 2.373 ± 0.013 | 2.356 | 2.395 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base pedersen` | 3.340 ± 0.010 | 3.328 | 3.365 | 1.00 ± 0.00 |
| `head pedersen` | 3.332 ± 0.006 | 3.325 | 3.344 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base poseidon_integration_benchmark` | 1.492 ± 0.003 | 1.487 | 1.498 | 1.00 |
| `head poseidon_integration_benchmark` | 1.499 ± 0.012 | 1.488 | 1.526 | 1.01 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base secp_integration_benchmark` | 2.621 ± 0.014 | 2.607 | 2.657 | 1.00 ± 0.01 |
| `head secp_integration_benchmark` | 2.620 ± 0.009 | 2.610 | 2.639 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base set_integration_benchmark` | 1.841 ± 0.018 | 1.819 | 1.875 | 1.00 ± 0.01 |
| `head set_integration_benchmark` | 1.838 ± 0.016 | 1.822 | 1.867 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base uint256_integration_benchmark` | 4.147 ± 0.018 | 4.108 | 4.171 | 1.00 |
| `head uint256_integration_benchmark` | 4.151 ± 0.008 | 4.137 | 4.163 | 1.00 ± 0.00 |

**Comment by Oppen** [comment]
Note: for simplicity, next experiments will only have the parameters used and the reported results. From those, the acceptance criteria will be:
- For each __benchmark__, the reported standard deviation will be `<= 2%`.
- For each __benchmark__, the max and min values will be within `mean +- 2 * stddev` (this one may change, some outliers may be acceptable within some range).
- For each __program__, the reported difference will be `<= 2%`.
- For each __program__, the reported uncertainty in the difference will be `<= 2%`.
- For each __program__, the reported uncertainty in the difference will be `>=` than the reported difference.

What this ensures is there's no significant outliers and that all difference is attributable to noise, and that such noise is not too big.
From acceptable runs we should pick the fastest to improve our turnaround times.

**Comment by Oppen** [comment]
Experiment 4: `cat {program} {bin}`, 2 cores, no warmup. Time: 8'52".

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base blake2s_integration_benchmark` | 2.791 ± 0.051 | 2.736 | 2.872 | 1.01 ± 0.02 |
| `head blake2s_integration_benchmark` | 2.774 ± 0.012 | 2.756 | 2.792 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base compare_arrays_200000` | 3.367 ± 0.042 | 3.325 | 3.438 | 1.01 ± 0.01 |
| `head compare_arrays_200000` | 3.340 ± 0.020 | 3.314 | 3.375 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base dict_integration_benchmark` | 2.533 ± 0.018 | 2.517 | 2.560 | 1.00 |
| `head dict_integration_benchmark` | 2.540 ± 0.033 | 2.505 | 2.625 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base factorial_multirun` | 4.093 ± 0.033 | 4.061 | 4.[16](https://github.com/lambdaclass/cairo-rs/actions/runs/4690495994/jobs/8313702016?pr=965#step:7:17)2 | 1.00 |
| `head factorial_multirun` | 4.104 ± 0.038 | 4.070 | 4.204 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base fibonacci_1000_multirun` | 3.268 ± 0.014 | 3.248 | 3.290 | 1.00 |
| `head fibonacci_1000_multirun` | 3.280 ± 0.027 | 3.250 | 3.342 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base integration_builtins` | 3.344 ± 0.019 | 3.325 | 3.387 | 1.00 |
| `head integration_builtins` | 3.354 ± 0.019 | 3.338 | 3.397 | 1.00 ± 0.01 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base keccak_integration_benchmark` | 3.031 ± 0.009 | 3.018 | 3.047 | 1.00 |
| `head keccak_integration_benchmark` | 3.051 ± 0.045 | 3.021 | 3.168 | 1.01 ± 0.02 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base linear_search` | 3.494 ± 0.033 | 3.469 | 3.569 | 1.00 ± 0.02 |
| `head linear_search` | 3.483 ± 0.054 | 3.449 | 3.634 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base math_cmp_and_pow_integration_benchmark` | 2.924 ± 0.034 | 2.896 | 3.0[17](https://github.com/lambdaclass/cairo-rs/actions/runs/4690495994/jobs/8313702016?pr=965#step:7:18) | 1.00 |
| `head math_cmp_and_pow_integration_benchmark` | 2.926 ± 0.048 | 2.891 | 3.017 | 1.00 ± 0.02 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base math_integration_benchmark` | 2.636 ± 0.009 | 2.627 | 2.656 | 1.00 ± 0.00 |
| `head math_integration_benchmark` | 2.629 ± 0.004 | 2.624 | 2.635 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base memory_integration_benchmark` | 2.337 ± 0.014 | 2.320 | 2.359 | 1.00 ± 0.02 |
| `head memory_integration_benchmark` | 2.335 ± 0.037 | 2.317 | 2.436 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base operations_with_data_structures_benchmarks` | 2.348 ± 0.015 | 2.331 | 2.375 | 1.00 ± 0.01 |
| `head operations_with_data_structures_benchmarks` | 2.340 ± 0.009 | 2.331 | 2.358 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base pedersen` | 3.331 ± 0.006 | 3.320 | 3.338 | 1.00 ± 0.00 |
| `head pedersen` | 3.324 ± 0.004 | 3.319 | 3.334 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base poseidon_integration_benchmark` | 1.473 ± 0.003 | 1.468 | 1.477 | 1.00 |
| `head poseidon_integration_benchmark` | 1.476 ± 0.006 | 1.470 | 1.490 | 1.00 ± 0.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base secp_integration_benchmark` | 2.615 ± 0.0[18](https://github.com/lambdaclass/cairo-rs/actions/runs/4690495994/jobs/8313702016?pr=965#step:7:19) | 2.597 | 2.662 | 1.00 ± 0.01 |
| `head secp_integration_benchmark` | 2.614 ± 0.015 | 2.601 | 2.647 | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base set_integration_benchmark` | 1.813 ± 0.024 | 1.795 | 1.879 | 1.00 ± 0.01 |
| `head set_integration_benchmark` | 1.812 ± 0.006 | 1.805 | 1.8[23](https://github.com/lambdaclass/cairo-rs/actions/runs/4690495994/jobs/8313702016?pr=965#step:7:24) | 1.00 |
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base uint[25](https://github.com/lambdaclass/cairo-rs/actions/runs/4690495994/jobs/8313702016?pr=965#step:7:26)6_integration_benchmark` | 4.113 ± 0.013 | 4.098 | 4.1[36](https://github.com/lambdaclass/cairo-rs/actions/runs/4690495994/jobs/8313702016?pr=965#step:7:37) | 1.00 ± 0.01 |
| `head uint256_integration_benchmark` | 4.105 ± 0.017 | 4.088 | 4.1[47](https://github.com/lambdaclass/cairo-rs/actions/runs/4690495994/jobs/8313702016?pr=965#step:7:48) | 1.00 |

**Comment by Oppen** [comment]
Experiment 4 seems to reach acceptable results while being faster. So I'll go with that unless I see any objections.

**Review by juanbono — APPROVED** [review]

File: .github/workflows/hyperfine.yml [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        with:
+          ref: ${{ github.event.pull_request[matrix.branch].sha }}
+
+      - name: Fetch from cache
+        uses: actions/cache@v3
+        id: cache
+        with:
+          path: base_programs/*.json
+          key: benchmarks-${{ hashFiles( 'cairo_programs/benchmarks/*.cairo' ) }}
+          restore-keys: benchmarks-
```
Comment (MegaRedHand) [inline .github/workflows/hyperfine.yml:36]: Wouldn't this, combined with skipping the steps of contract compilation on cache hit, produce errors when one of them changed?

**Review by MegaRedHand — APPROVED** [review]

**Pull request merged**